Install Pytorch package (need to specify the version to get correctly installed. Possible error will lead died kernal.
)

In [1]:
# pip install torch==1.11.0 torchvision==0.12.0 torchaudio==0.11.0

The RNN model with PyTorch will be implemented on the Metadata only to farmiliarize myself with the PyTorch library and get starts with RNN. First of all, we need to treat each patient as an observation and get every 

In [2]:
import pandas as pd
import numpy as np

import torch
import torch.nn as nn

from sklearn.metrics import accuracy_score

A better explanation about the RNN example https://blog.floydhub.com/a-beginners-guide-on-recurrent-neural-networks-with-pytorch/

Read in data set

In [3]:
mydata = pd.read_csv('test/metadata_imputed.csv', delimiter=',')
mydata = pd.DataFrame(mydata)
mydata

participant_id project   specimen  was_term  delivery_wk  collect_wk  \
0            A00001       A  A00001-05      True           38          33   
1            A00002       A  A00002-01      True           40          38   
2            A00003       A  A00003-02      True           40          30   
3            A00004       A  A00004-08      True           40          27   
4            A00004       A  A00004-12      True           40          29   
...             ...     ...        ...       ...          ...         ...   
3573         J00111       J  J00111-01      True           40          17   
3574         J00112       J  J00112-01      True           39          19   
3575         J00113       J  J00113-01      True           41          16   
3576         J00115       J  J00115-01      True           42          18   
3577         J00116       J  J00116-01      True           40          17   

                 race      age               NIH.Racial.Category  \
0     American Indian  Unknown  American Indian or Alaska Native   
1               White  Unknown                             White   
2      Asian-Japanese  Unknown                             Asian   
3               White  Unknown                             White   
4               White  Unknown                             White   
...               ...      ...                               ...   
3573        Caucasian       27                             White   
3574        Caucasian       27                             White   
3575        Caucasian       32                             White   
3576        Caucasian       35                             White   
3577        Caucasian       26                             White   

     NIH.Ethnicity.Category  was_preterm  was_early_preterm  collect_tri  \
0                   Unknown        False              False            3   
1                   Unknown        False              False            3   
2                   Unknown        False              False            3   
3                   Unknown        False              False            3   
4                   Unknown        False              False            3   
...                     ...          ...                ...          ...   
3573                Unknown        False              False            2   
3574                Unknown        False              False            2   
3575                Unknown        False              False            2   
3576                Unknown        False              False            2   
3577                Unknown        False              False            2   

      age_imp                          race_imp    age_imp_cat  
0          27  American Indian or Alaska Native  from_18_to_28  
1          24                             White  from_18_to_28  
2          32                             Asian  from_29_to_38  
3          25                             White  from_18_to_28  
4          25                             White  from_18_to_28  
...       ...                               ...            ...  
3573       27                             White  from_18_to_28  
3574       27                             White  from_18_to_28  
3575       32                             White  from_29_to_38  
3576       35                             White  from_29_to_38  
3577       26                             White  from_18_to_28  

[3578 rows x 16 columns]

In [4]:
mydata = mydata[["participant_id", "project", "specimen", "collect_wk", 
                 "was_preterm", "was_early_preterm", "race_imp", "age_imp_cat" ]]
mydata["project"] = mydata["project"].astype('category')
mydata["race_imp"] = mydata["race_imp"].astype('category')
mydata["age_imp_cat"] = mydata["age_imp_cat"].astype('category')

mydata["project"] = mydata["project"].cat.codes
mydata["race_imp"] = mydata["race_imp"].cat.codes
mydata["age_imp_cat"] = mydata["age_imp_cat"].cat.codes
mydata['was_preterm'] = mydata['was_preterm'].astype('int8')
mydata['was_early_preterm'] = mydata['was_early_preterm'].astype('int8')
mydata.dtypes

C:\Users\sxiao15\AppData\Local\Temp\ipykernel_15180\1563144068.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mydata["project"] = mydata["project"].astype('category')
C:\Users\sxiao15\AppData\Local\Temp\ipykernel_15180\1563144068.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mydata["race_imp"] = mydata["race_imp"].astype('category')
C:\Users\sxiao15\AppData\Local\Temp\ipykernel_15180\1563144068.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try u

participant_id       object
project                int8
specimen             object
collect_wk            int64
was_preterm            int8
was_early_preterm      int8
race_imp               int8
age_imp_cat            int8
dtype: object

We split the training and testing data set based on *project* J.

In [5]:
mydata_train = mydata[mydata['project']!=9] # Not project J
mydata_test  = mydata[mydata['project']==9] # project J

In [6]:
# mydata_train = mydata_train[["participant_id", "project", "specimen", "collect_wk", 
#                              "was_preterm", "was_early_preterm", "race_imp", "age_imp_cat" ]]

In [7]:
# mydata1["project"] = mydata1["project"].astype('category')
# mydata1["race_imp"] = mydata1["race_imp"].astype('category')
# mydata1["age_imp_cat"] = mydata1["age_imp_cat"].astype('category')

In [8]:
# mydata1["project"] = mydata1["project"].cat.codes
# mydata1["race_imp"] = mydata1["race_imp"].cat.codes
# mydata1["age_imp_cat"] = mydata1["age_imp_cat"].cat.codes
# mydata1['was_preterm'] = mydata1['was_preterm'].astype('int8')
# mydata1['was_early_preterm'] = mydata1['was_early_preterm'].astype('int8')
# mydata1.dtypes

We want to predict the class of preterm birth and early preterm birth for few consecutive specimemns collected. We will have to define the

- Input data (2 covariates)
    - The meta data (imputed age and race category) for each observation
- Output (1 covariate)
    - The class label of preterm birth/early preterm birth
    
via data pre-processing and move the data from numpy arrays to Pytorch data structure - **Torch Tensors**

## Preterm birth

Since the simple RNN only allow one covariates to enter the model, I choose *age_imp_cat* as a starting point.

### Input data from categorical age

We need to do some data filtering.

In [9]:
# only keep the rows with collect_wk < 37 for Preterm task
mydata1 = mydata_train.loc[mydata_train['collect_wk']<37,]

# delete the duplicated samples in the same collection week
mydata2 = mydata1.drop_duplicates(subset=['participant_id', 'collect_wk'], keep='first')
mydata2.shape

(2417, 8)

Note: Too much samples are deleted, we need to take care of this situation. For metadata table, its OK since the information of duplicates are exactly same. But it will be specimen specific for taxonomy, phylotypes and distance data sets. **We need to discuss it**. We need to either form up some summary statistic for multiple collected sample in the same week or other methods (???).

Since we will give a patient the label of preterm birth as the delievey week < 37, we have 36 as one of the dimension for the tensor.

In [10]:
mydata2_wide_age = mydata2.pivot_table(index=['participant_id'], columns='collect_wk', values='age_imp_cat')
# sort by collect_wk
mydata2_wide_age = mydata2_wide_age.sort_index(axis=1)
mydata2_wide_age = mydata2_wide_age.fillna(0)
mydata2_wide_age.shape
mydata2_wide_age.head

<bound method NDFrame.head of collect_wk       1    2    3    4    5    6    7    8    9    10  ...   27  \
participant_id                                                    ...        
A00001          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
A00003          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
A00004          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  2.0   
A00005          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
A00006          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
...             ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
I00534          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
I00535          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
I00536          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
I00537          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
I00538          0.0  0.0  0.0  0.0

In [11]:
(seq_length, num_feature) = mydata2_wide_age.shape
print(seq_length)
print(num_feature)

1174
36


### Output data from Preterm birth label

In [12]:
mydata2_wide_Y = mydata2.pivot_table(index=['participant_id'], columns='collect_wk', values='was_preterm')
# sort by collect_wk
mydata2_wide_Y = mydata2_wide_Y.sort_index(axis=1)
mydata2_wide_Y = mydata2_wide_Y.apply(lambda row: row.fillna(row.mean()), axis=1)
mydata2_wide_Y.head(100)

collect_wk       1    2    3    4    5    6    7    8    9    10  ...   27  \
participant_id                                                    ...        
A00001          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
A00003          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
A00004          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
A00005          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
A00006          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
...             ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
C00026          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0   
C00027          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0   
C00028          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0   
C00029          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0   
C00030          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0   

collect_wk       28   29   30   31   32   33   34   35   36  
participant_id                                               
A00001          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
A00003          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
A00004          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
A00005          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
A00006          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
...             ...  ...  ...  ...  ...  ...  ...  ...  ...  
C00026          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  
C00027          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  
C00028          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  
C00029          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  
C00030          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  

[100 rows x 36 columns]

In [13]:
# Creating lists that will hold our input and target sequences
input_seq = torch.Tensor([mydata2_wide_age.to_numpy()])
target_seq = torch.Tensor([mydata2_wide_Y.to_numpy()])

C:\Users\sxiao15\AppData\Local\Temp\ipykernel_15180\1327531968.py:2: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\torch\csrc\utils\tensor_new.cpp:210.)
  input_seq = torch.Tensor([mydata2_wide_age.to_numpy()])


In [14]:
input_seq.shape

torch.Size([1, 1174, 36])

In [15]:
target_seq.shape

torch.Size([1, 1174, 36])

In [16]:
# torch.cuda.is_available() checks and returns a Boolean True if a GPU is available, else it'll return False
is_cuda = torch.cuda.is_available()

# If we have a GPU available, we'll set our device to GPU. We'll use this device variable later in our code.
if is_cuda:
    device = torch.device("cuda")
    print("GPU is available")
else:
    device = torch.device("cpu")
    print("GPU not available, CPU used")

GPU not available, CPU used


In [17]:
class Model(nn.Module):
    def __init__(self, input_size, output_size, hidden_dim, n_layers):
        super(Model, self).__init__()

        # Defining some parameters
        self.hidden_dim = hidden_dim
        self.n_layers = n_layers

        # Defining the layers
        # RNN Layer
        self.rnn = nn.RNN(input_size, hidden_dim, n_layers, batch_first=True, nonlinearity='relu')   
        # Fully connected layer
        self.fc = nn.Linear(hidden_dim, output_size)
    
    def forward(self, x):
        
        batch_size = x.size(0)

        #Initializing hidden state for first input using method defined below
        hidden = self.init_hidden(batch_size)

        # Passing in the input and hidden state into the model and obtaining outputs
        out, hidden = self.rnn(x, hidden)
        
        # Reshaping the outputs such that it can be fit into the fully connected layer
        out = out.contiguous().view(-1, self.hidden_dim)
        out = self.fc(out)
        
        return out, hidden
    
    def init_hidden(self, batch_size):
        # This method generates the first hidden state of zeros which we'll use in the forward pass
        hidden = torch.zeros(self.n_layers, batch_size, self.hidden_dim).to(device)
         # We'll send the tensor holding the hidden state to the device we specified earlier as well
        return hidden

In [18]:
# Instantiate the model with hyperparameters
model = Model(input_size=num_feature, output_size=num_feature, hidden_dim=12, n_layers=3)
# We'll also set the model to the device that we defined earlier (default is CPU)
model = model.to(device)

# Define hyperparameters
n_epochs = 100
lr=0.01

# Define Loss, Optimizer
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=lr)

In [19]:
# Training Run
input_seq = input_seq.to(device)
for epoch in range(1, n_epochs + 1):
    optimizer.zero_grad() # Clears existing gradients from previous epoch
    #input_seq = input_seq.to(device)
    output, hidden = model(input_seq)
    output = output.to(device)
    target_seq = target_seq.to(device)
    loss = criterion(output, target_seq.view(-1, num_feature))
    loss.backward() # Does backpropagation and calculates gradients
    optimizer.step() # Updates the weights accordingly
    
    if epoch%10 == 0:
        print('Epoch: {}/{}.............'.format(epoch, n_epochs), end=' ')
        print("Loss: {:.4f}".format(loss.item()))

Epoch: 10/100............. Loss: 44.2048
Epoch: 20/100............. Loss: 44.1793
Epoch: 30/100............. Loss: 44.1754
Epoch: 40/100............. Loss: 44.1751
Epoch: 50/100............. Loss: 44.1745
Epoch: 60/100............. Loss: 44.1744
Epoch: 70/100............. Loss: 44.1744
Epoch: 80/100............. Loss: 44.1744
Epoch: 90/100............. Loss: 44.1744
Epoch: 100/100............. Loss: 44.1744


In [20]:
target_seq.view(-1, num_feature).shape

torch.Size([1174, 36])

### Apply the simple RNN on testing set.

In [21]:
testset = mydata_test[["participant_id", "project", "specimen", "collect_wk", 
                       "was_preterm", "was_early_preterm", "race_imp", "age_imp_cat" ]]
print(testset.dtypes)
print(testset.shape)

participant_id       object
project                int8
specimen             object
collect_wk            int64
was_preterm            int8
was_early_preterm      int8
race_imp               int8
age_imp_cat            int8
dtype: object
(83, 8)


In [22]:
# only keep the rows with collect_wk < 37 for Preterm task
testset1 = testset.loc[testset['collect_wk']<37,]

# delete the duplicated samples in the same collection week
testset2 = testset1.drop_duplicates(subset=['participant_id', 'collect_wk'], keep='first')
testset2.shape

(83, 8)

In [23]:
testset2_wide_age = testset2.pivot_table(index=['participant_id'], columns='collect_wk', values='age_imp_cat')
# sort by collect_wk
testset2_wide_age = testset2_wide_age.sort_index(axis=1)
testset2_wide_age = testset2_wide_age.fillna(0)
testset2_wide_age.shape
testset2_wide_age.head

<bound method NDFrame.head of collect_wk       12   13   14   15   16   17   18   19
participant_id                                        
J00001          0.0  0.0  0.0  0.0  0.0  0.0  3.0  0.0
J00004          0.0  0.0  0.0  0.0  0.0  2.0  0.0  0.0
J00007          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
J00008          0.0  0.0  0.0  0.0  0.0  0.0  3.0  0.0
J00010          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
...             ...  ...  ...  ...  ...  ...  ...  ...
J00111          0.0  0.0  0.0  0.0  0.0  2.0  0.0  0.0
J00112          0.0  0.0  0.0  0.0  0.0  0.0  0.0  2.0
J00113          0.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0
J00115          0.0  0.0  0.0  0.0  0.0  0.0  3.0  0.0
J00116          0.0  0.0  0.0  0.0  0.0  2.0  0.0  0.0

[83 rows x 8 columns]>

Fill-in zeros for empty collection weeks.

In [24]:
(seq_length_test, num_feature_test) = testset2_wide_age.shape

In [25]:
testset3 = pd.DataFrame(np.nan, index=range(seq_length_test), columns=range(num_feature))
testset3 = testset3.rename(columns={x:y for x,y in zip(testset3.columns,range(0,num_feature))})
testset3 = testset3.fillna(0)
print(testset3.shape)
test_index = testset2_wide_age._get_numeric_data().columns.values.tolist()
testset3.iloc[:,test_index] = testset2_wide_age

# check: first observation in testing set has value 3.0 at column 19 (index 18)
testset3.iloc[0]

(83, 36)


0     0.0
1     0.0
2     0.0
3     0.0
4     0.0
5     0.0
6     0.0
7     0.0
8     0.0
9     0.0
10    0.0
11    0.0
12    0.0
13    0.0
14    0.0
15    0.0
16    0.0
17    0.0
18    3.0
19    0.0
20    0.0
21    0.0
22    0.0
23    0.0
24    0.0
25    0.0
26    0.0
27    0.0
28    0.0
29    0.0
30    0.0
31    0.0
32    0.0
33    0.0
34    0.0
35    0.0
Name: 0, dtype: float64

In [26]:
ii = 0
mytest_input = testset3.iloc[ii]
mytest_input = torch.Tensor([[mytest_input.to_numpy()]])
mytest_input.shape

torch.Size([1, 1, 36])

In [27]:
model.eval() # eval mode
out, hidden = model(mytest_input)
prob = nn.functional.softmax(out[-1], dim=0).data
# Taking the class with the highest probability score from the output
prob_max = torch.max(prob)
print(prob_max)
pred_label = 1*(prob_max > 0.5)
pred_label

tensor(0.0281)


tensor(0)

In [28]:
model.eval() # eval mode

predicted_labels = []
for ii in range(seq_length_test):
    mytest_input = testset3.iloc[ii]
    mytest_input = torch.Tensor([[mytest_input.to_numpy()]])
    out, hidden = model(mytest_input)
    prob = nn.functional.softmax(out[-1], dim=0).data
    # Taking the class with the highest probability score from the output
    prob_max = torch.max(prob).item()
    print(prob_max)
    pred_label = 1*(prob_max > 0.5)
    predicted_labels.append(pred_label)

0.028099045157432556
0.028104817494750023
0.028104789555072784
0.028099045157432556
0.028104789555072784
0.02810567058622837
0.028099045157432556
0.02809915877878666
0.02808426134288311
0.028104793280363083
0.028104789555072784
0.028092756867408752
0.028104789555072784
0.02810567058622837
0.02810567058622837
0.028099045157432556
0.028104789555072784
0.028104793280363083
0.028104789555072784
0.02810567058622837
0.028104789555072784
0.02808426134288311
0.02809915877878666
0.02809915877878666
0.02809915877878666
0.02809915877878666
0.028104793280363083
0.028104789555072784
0.02810567058622837
0.02810567058622837
0.02810567058622837
0.028104793280363083
0.02810577303171158
0.028099045157432556
0.02810567058622837
0.028104793280363083
0.02810567058622837
0.028098834678530693
0.028104793280363083
0.02810567058622837
0.028092706575989723
0.028104817494750023
0.02810567058622837
0.02810567058622837
0.02810148522257805
0.02810567058622837
0.028104817494750023
0.028101855888962746
0.028104789555

In [29]:
predicted_labels

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0]

Note: since we only involved *age_imp_cat* in this baby RNN model, all patients in project J are poredicted to not have preterm with a very small p-value.

In [30]:
testset2_wide_Y = testset2.pivot_table(index=['participant_id'], columns='collect_wk', values='was_preterm')
# sort by collect_wk
testset2_wide_Y = testset2_wide_Y.sort_index(axis=1)
actual_labels = testset2_wide_Y.mean(axis=1)
print(actual_labels.iloc[1:80])

participant_id
J00004    0.0
J00007    0.0
J00008    0.0
J00010    0.0
J00011    0.0
         ... 
J00107    0.0
J00108    0.0
J00109    0.0
J00111    0.0
J00112    0.0
Length: 79, dtype: float64


In [31]:
# create vectors for actual labels and predicted labels...
my_accuracy = accuracy_score(actual_labels, predicted_labels, normalize=False) / float(actual_labels.size)
my_accuracy 

0.8192771084337349

In [38]:
# Creating lists that will hold our input and target sequences
input_seq = torch.Tensor([mydata2_wide_age.to_numpy()]).view(mydata2_wide_age.to_numpy().shape[1],-1,1)
target_seq = torch.Tensor([np.vstack((mydata2_wide_Y.to_numpy().flatten(),
                                    1-mydata2_wide_Y.to_numpy().flatten())).T]).view(-1,2)

print(input_seq.shape)
print(target_seq.shape)

torch.Size([36, 1174, 1])
torch.Size([42264, 2])


### GRU model

In [39]:
# Define GRU model
class GRUModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim, n_layers = 1,drop_prob=0.2):
        
        """
            parameters:
                input_dim: dimensions of input data (# features)
                hidden_dim: dimensions of hidden layer
                output_dim: dimensions of output layer (should be two in our analysis)
                n_layers: number of layers for GRU structure, default is 1, 2 means stacked GRU
                drop_prob: dropout probability

        """
        
        #inherit from super class
        super(GRUModel, self).__init__()
        
        #define parameters
        self.hidden_dim = hidden_dim
        self.n_layers = n_layers
        
        #define layers
        
        ##GRU layers
        self.gru = nn.GRU(input_dim, hidden_dim, n_layers, dropout=drop_prob)
        
        ##fully connected layer(use one linear layer first, later can customize)
        self.fc = nn.Linear(hidden_dim, output_dim)
        
    def forward(self, x):
        
        batch_size = x.size(1)

        #Initializing hidden state
        hidden = self.init_hidden(batch_size)
        
        out, hidden = self.gru(x, hidden)
        
        print(out.shape)
        
        #pass out to fully connected layer
        out = self.fc(out.reshape(-1,out.shape[-1]))
        
        return out, hidden
    
    def init_hidden(self, batch_size):
        return torch.zeros((self.n_layers,batch_size, self.hidden_dim), device = device)
    
    
    



In [34]:
# Define LSTM model

class LSTMModel(nn.Module):
    
    """
        parameters:
            input_dim: dimensions of input data (# features)
            hidden_dim: dimensions of hidden layer
            output_dim: dimensions of output layer (should be two in our analysis)
            n_layers: number of layers for GRU structure, default is 1, 2 means stacked GRU
            drop_prob: dropout probability

    """
        
    def __init__(self, input_dim, hidden_dim, output_dim, n_layers = 1, drop_prob=0.2):
        
        #inherit from super class
        super(LSTMModel, self).__init__()
        
        #define parameters
        self.hidden_dim = hidden_dim
        self.n_layers = n_layers
        
        #define layers
        
        ##LSTM layers
        self.lstm = nn.LSTM(input_dim, hidden_dim, n_layers, dropout=drop_prob)
        
        ##fully connected layer(use one linear layer first, later can customize)
        self.fc = nn.Linear(hidden_dim, output_dim)
        
        
    def forward(self, x):
        
        batch_size = x.size(1)

        #Initializing hidden state
        hidden = self.init_hidden(batch_size)
        
        out, hidden = self.lstm(x, hidden)
        
        #pass out to fully connected layer
        out = self.fc(out.reshape(-1,out.shape[-1]))
        
        return out, hidden
    
    def init_hidden(self, batch_size):
        return (torch.zeros((self.n_layers,batch_size, self.hidden_dim),device = device),
                torch.zeros((self.n_layers,batch_size, self.hidden_dim),device = device))

In [40]:
# define train function


def train_epoch(X_train, y_train, hidden_dim, learn_rate, device, EPOCHS,output_dim = 2, method = "GRU"):
    
    input_dim = X_train.shape[2]
    
    print(input_dim)
    
    
    if method == "GRU":
        model = GRUModel(input_dim, hidden_dim, output_dim)
    elif method == "LSTM":
        model = LSTMModel(input_dim, hidden_dim, output_dim)
        
    model.to(device)
    
    # loss criterion and optimizer

    criterion = nn.CrossEntropyLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr = learn_rate)

    
    # train model
    
    model.train()
    
    print('Starting training of {} model'.format(method))
    

    X_train = X_train.to(device)
    y_train = y_train.to(device)
    
    #Start training loop
    
    for epoch in range(1, EPOCHS + 1):
        
        
        # Clears existing gradients from previous epoch
        optimizer.zero_grad() 
        
        predictions = model(X_train)[0]
        
        predictions = predictions.to(device)
        
        loss = criterion(predictions, y_train)
        # backpropagation
        loss.backward() 
        # Updates the weights accordingly
        optimizer.step()
    
        if epoch%10 == 0:
            print('Epoch: {}/{}.............'.format(epoch, EPOCHS), end=' ')
            print("Loss: {:.4f}".format(loss.item()))
        
    return model


In [41]:
GRU_model = train_epoch(X_train = input_seq, y_train = target_seq, hidden_dim = 12, learn_rate = 0.01, device = device, EPOCHS = 100, output_dim = 2, method = "GRU")

1
Starting training of GRU model
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
Epoch: 10/100............. Loss: 0.6407
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
Epoch: 20/100............. Loss: 0.6300
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
torch.Size([36, 1174, 12])
Epoch: 30/100............. Loss: 0.6017
torch.Size([36, 1174, 12])
torch.Size

In [37]:
LSTM_model = train_epoch(X_train = input_seq, y_train = target_seq, hidden_dim = 12, learn_rate = 0.01, device = device, EPOCHS = 100, output_dim = 2, method = "LSTM")

1
Starting training of LSTM model
Epoch: 10/100............. Loss: 0.6440
Epoch: 20/100............. Loss: 0.6393
Epoch: 30/100............. Loss: 0.6159
Epoch: 40/100............. Loss: 0.5922
Epoch: 50/100............. Loss: 0.5811
Epoch: 60/100............. Loss: 0.5748
Epoch: 70/100............. Loss: 0.5710
Epoch: 80/100............. Loss: 0.5649
Epoch: 90/100............. Loss: 0.5587
Epoch: 100/100............. Loss: 0.5545


Questions to discuss:

1. How many outputs are we expected? I think per my reading, there should be one output for each timepoint. In our case, there should be a preterm/non-preterm outcome at each timepoint right? In the literatures I read these days, I found most of the usage of RNN or modified RNN was in text prediction or longitudinal prediction, where there were different outcomes at each timepoint(or sequence position), but our outcome was kind of static. I was wondering if RNN would work for static outcome. 


2. After checking Mo's code and pytorch RNN tutorial(GRU and LSTM are similar to RNN), the input data should be in the shape of (N,L,H), where N is the batch size, L is sequence length(time point) and H is the dimension of input features. They should be 1174, 36 and 1 in our toy example with only age, but Mo's input shape is (1,1174,36), which confused me, but it seemed to work as well. 